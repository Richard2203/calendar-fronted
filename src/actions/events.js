import Swal from 'sweetalert2';
import { types } from '../types/types';
import { fetchConToken } from '../helpers/fetch';
import { Events_Parse } from '../helpers/Events_Parse';

export const eventStartAddNew = (event) => {
	return async (dispatch, getState) => {
		const { name, uid } = getState().auth;

		try {
			const resp = await fetchConToken('events', event, 'POST');
			const body = await resp.json();

			if (body.ok) {
				// agregando la propieadad "id" al objeto "event"
				event.id = body.evento.id;

				// agregando la propiedad "user" al objeto "event"
				event.user = { _id: uid, name };

				// guardando el "event" en Redux
				dispatch(eventAddNew(event));
			}
		} catch (error) {
			console.log(error);
		}
	};
};

const eventAddNew = (event) => ({
	type: types.eventAddNew,
	payload: event,
});

export const eventSetActive = (event) => ({
	type: types.eventSetActive,
	payload: event,
});

export const eventClearActiveEvent = () => ({
	type: types.eventClearActiveEvent,
});

export const eventStartUpdated = (event) => {
	return async (dispatch) => {
		try {
			const resp = await fetchConToken(
				`events/${event.id}`,
				event,
				'PUT'
			);
			const body = await resp.json();

			if (body.ok) dispatch(eventUpdated(event));
			else Swal.fire('error', body.msg, 'error');
		} catch (error) {
			console.log(error);
		}
	};
};

const eventUpdated = (event) => ({
	type: types.eventUpdated,
	payload: event,
});

export const eventStartDelete = () => {
	return async (dispatch, getState) => {
		const { id } = getState().calendar.activeEvent;
		try {
			const resp = await fetchConToken(`events/${id}`, {}, 'DELETE');

			const body = await resp.json();

			if (body.ok) dispatch(eventDeleted());
			else Swal.fire('error', body.msg, 'error');
		} catch (error) {
			console.log(error);
		}
	};
};

const eventDeleted = () => ({
	type: types.eventDeleted,
});

export const eventStartLoading = () => {
	return async (dispatch) => {
		try {
			const resp = await fetchConToken('events');
			const { events } = await resp.json();

			// obteniendo los events retornados de la peticion fetch
			const ArrayEvents = Events_Parse(events);

			dispatch(eventLoaded(ArrayEvents));
		} catch (error) {
			console.log(error);
		}
	};
};

const eventLoaded = (events) => ({ type: types.eventLoaded, payload: events });

export const Logout = () => ({ type: types.eventLogout });
