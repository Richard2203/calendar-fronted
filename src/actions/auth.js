import Swal from 'sweetalert2';
import { fetchConToken, fetchSinToken } from '../helpers/fetch';
import { types } from '../types/types';
import { Logout } from './events';

export const startLogin = (userCreden) => {
	return async (dispatch) => {
		const res = await fetchSinToken('auth', userCreden, 'POST');
		const { ok, token, uid, name, msg } = await res.json();

		if (ok) {
			localStorage.setItem('token', token);
			localStorage.setItem('token-init-date', new Date().getTime());
			dispatch(login({ uid, name }));
		} else Swal.fire('Error', msg, 'error');
	};
};

export const startRegister = (name, email, password) => {
	return async (dispatch) => {
		const res = await fetchSinToken(
			'auth/new',
			{ name, email, password },
			'POST'
		);

		const { ok, name: nam, uid, token, msg } = await res.json();

		if (ok) {
			localStorage.setItem('token', token);
			localStorage.setItem('token-init-date', new Date().getTime());
			dispatch(login({ nam, uid }));
		} else Swal.fire('Error', msg, 'error');
	};
};

export const startCheking = () => {
	return async (dispatch) => {
		if (!localStorage.getItem('token')) {
			dispatch(checkingFinish());
			return;
		}

		const res = await fetchConToken('auth/renew');

		const { ok, name, uid, token } = await res.json();

		if (ok) {
			localStorage.setItem('token', token);
			localStorage.setItem('token-init-date', new Date().getTime());
			dispatch(login({ name, uid }));
		} else dispatch(checkingFinish());
	};
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

export const startLogout = () => {
	return (dispatch) => {
		// implementando el localStorage en un action puesto que de ponerla
		// directamente en el reducer puede marcar error. La logica se debe
		// separar

		// purga todo el localStorage
		localStorage.clear();

		// aqui se purga la informacion en el reducer
		dispatch(authLogout());

		dispatch(Logout());
	};
};

const authLogout = () => ({ type: types.authLogout });

const login = (user) => ({
	type: types.authLogin,
	payload: user,
});
