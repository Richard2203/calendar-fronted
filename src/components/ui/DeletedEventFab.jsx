import React from 'react';
import { useDispatch } from 'react-redux';
import { eventDeleted } from '../../actions/events';
import './FabStyles.css';

const DeletedEventFab = () => {
	const dispatch = useDispatch();

	const handleDelete = () => {
		dispatch(eventDeleted());
	};

	return (
		<button
			className="btn btn-outline-danger fab-danger"
			onClick={handleDelete}
		>
			<i className="fa fa-trash"></i>
			<span>Borrar Evento</span>
		</button>
	);
};

export default DeletedEventFab;
