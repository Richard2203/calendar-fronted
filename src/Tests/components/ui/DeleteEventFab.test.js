import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import DeletedEventFab from '../../../components/ui/DeletedEventFab';
import { eventStartDelete } from '../../../actions/events';

jest.mock('../../../actions/events', () => ({
	eventStartDelete: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Pruebas en <DeletedEventFab/>', () => {
	const wrapper = mount(
		<Provider store={store}>
			<DeletedEventFab />
		</Provider>
	);

	test('debe renderizar correctamente el componente', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('debe disparar eventStartDelete', () => {
		wrapper.find('button').prop('onClick')();

		// el dispatch debio ser llamado con una funcion
		expect(store.dispatch).toHaveBeenCalled();

		// debido a que creere un mock del argumento "eventStartDelete"
		// ya no es posible saber si dispatch fue llamado con ese argumento
		// expect(store.dispatch).toHaveBeenCalledWith(expect.any(Function));

		// eventStartDelete debio ser llamado
		expect(eventStartDelete).toHaveBeenCalled();
	});
});
