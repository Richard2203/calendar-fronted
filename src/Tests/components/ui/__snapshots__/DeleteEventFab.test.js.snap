// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <DeletedEventFab/> debe renderizar correctamente el componente 1`] = `
<button
  className="btn btn-outline-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fa fa-trash"
  />
  <span>
    Borrar Evento
  </span>
</button>
`;
